struct.go

```diff
@@ -74,22 +74,21 @@ type RegisterInstanceJson struct {
 	Instance *Instance `json:"instance"`
 }
 
-// Instance [de]serializeable [to|from] Eureka XML.
+// Instance [de]serializeable [to|from] Eureka [XML|JSON].
 type Instance struct {
-	XMLName          struct{} `xml:"instance" json:"-"`
-	HostName         string   `xml:"hostName" json:"hostName"`
-	App              string   `xml:"app" json:"app"`
-	IPAddr           string   `xml:"ipAddr" json:"ipAddr"`
-	VipAddress       string   `xml:"vipAddress" json:"vipAddress"`
-	SecureVipAddress string   `xml:"secureVipAddress" json:"secureVipAddress"`
+	HostName         string `xml:"hostName" json:"hostName"`
+	App              string `xml:"app" json:"app"`
+	IPAddr           string `xml:"ipAddr" json:"ipAddr"`
+	VipAddress       string `xml:"vipAddress" json:"vipAddress"`
+	SecureVipAddress string `xml:"secureVipAddress" json:"secureVipAddress"`
 
 	Status           StatusType `xml:"status" json:"status"`
 	Overriddenstatus StatusType `xml:"overriddenstatus" json:"overriddenstatus"`
 
-	Port        int  `xml:"port" json:"-"`
-	PortJ       Port `json:"port" xml:"-"`
-	SecurePort  int  `xml:"securePort" json:"-"`
-	SecurePortJ Port `json:"securePort" xml:"-"`
+	Port              int  `xml:"-" json:"-"`
+	PortEnabled       bool `xml:"-" json:"-"`
```

Comment:
I thought I had written about this earlier, but apparently not. I apologize for the delay in addressing your concern.

The message for commit 9352b649e800ab27140b0683283ddf16f6379f6e in this patch series mentions this removal:
> None of these were exposed deliberately for use by callers; they were
all exposed only as accidental consequences of the JSON and XML
marshalling library's idiosyncrasies.

We can't be sure whether anyone wrote to or read the "PortJ" and "SecurePortJ" fields. Removing them introduces an incompatibility, in the sense that code that used to compile will now fail to compile. I could add the fields back in, though the values read from those fields would be different than they were before, and writing to those fields wouldn't have the intended effect.

I had first tried to solve this problem by making use of those fields as they are, transferring values between them and making guesses as to whether setting a port to a nonzero value meant that it should be enabled, but I gave up on the idea and backed out those changes. As I wrote tests, I found that I couldn't even convince myself reliably of how it should work. It felt like we were going to just barely make it possible to register or detect which ports are enabled, but leave it too easy to get it wrong.

I am not sure that it's any more fair to our clients to allow their code to continue to compile as it did, but not work properly at run time. One thing I considered doing—and will do, if you think it's better than not doing it—is to leave those fields there, and when we _read_ an `Instance` from the Eureka server to transcribe the port number and whether it's enabled into those 'J' fields, duplicating the information from the "Port," "PortEnabled," "SecurePort," and "SecurePortEnabled" fields. _However_, I would not want to try something different in reverse: namely, I do not want to try to merge values from these different fields when marshaling an `Instance` to send to the Eureka server. To do so introduces too much confusion for callers: What happens if the values conflict? Which one do I need to set? Do I need to set both?

I doubted that _fargo_ had any clients accessing these fields profitably, but [I did find one public use](https://github.com/search?q=PortJ+SecurePortJ+language%3Ago&type=Code) over in [_stv3v/cfkit_](https://github.com/st3v/cfkit). There is also [_st3v/pcf-micro-examples_](https://github.com/st3v/pcf-micro-examples), but then I realized [that one is a vendored copy of _fargo_](https://github.com/st3v/pcf-micro-examples/tree/master/vendor/github.com/hudl/fargo). The former has [a legitimate use](https://github.com/st3v/cfkit/blob/4a389fa8fa8ca909ba57ede09054cfad6b68fdf7/discovery/eureka/eureka.go#L63-L64). @st3v is the author of [the commit that introduced those lines](https://github.com/st3v/cfkit/commit/707300ad4bacffce9bd86ae45ecf3a1d5ffe5303) on 1 October 2015. (Note that that setting of the "PortJ" and "SecurePortJ" fields would not have any effect if [`EurekaConnection.UseJson` had been false instead](https://github.com/st3v/cfkit/blob/4a389fa8fa8ca909ba57ede09054cfad6b68fdf7/discovery/eureka/eureka.go#L56).)

The _st3v/cfkit_ project has _fargo_'s version [pinned via Godeps](https://github.com/st3v/cfkit/blob/4a389fa8fa8ca909ba57ede09054cfad6b68fdf7/Godeps/Godeps.json#L32-L36) at commit ac37d05152fd52c1ba88771c908d8f6e6f1f340, which is from 10 September 2015. Hence, that project wouldn't be broken by this change unless it tried to advance its dependency version without changing its calling code.

@st3v Do you have an opinion on this matter?